backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/log/LogsRecomputationService.java

```diff
@@ -0,0 +1,211 @@
+			}
+
+			if (!logRecordRef.equals(currentRecordReference) || !logsIterator.hasNext())
+			{
+				recomputeForRecord(currentRecordReference);
+				currentRecordReference = logRecordRef;
+			}
+		}
+	}
+
+	public void recomputeForInvoice(@NonNull final IQueryFilter<I_C_Invoice> filter)
+	{
+		invoiceDAO.stream(filter)
+				.map(I_C_Invoice::getC_Invoice_ID)
+				.map(InvoiceId::ofRepoId)
+				.forEach(this::recomputeForInvoice);
+	}
+
+	public void recomputeForInOut(@NonNull final IQueryFilter<I_M_InOut> filter)
+	{
+		inOutDAO.stream(filter)
+				.forEach(this::recomputeForInOut);
+	}
+
+	public void recomputeForOrder(@NonNull final IQueryFilter<I_C_Order> filter)
+	{
+		orderDAO.streamOrders(filter)
+				.forEach(this::recomputeForOrder);
+	}
+
+	public void recomputeForFlatrate(@NonNull final IQueryFilter<I_C_Flatrate_Term> filter)
+	{
+		flatrateDAO.stream(filter)
+				.forEach(this::recomputeForFlatrate);
+	}
+
+	public void recomputeForInventory(@NonNull final IQueryFilter<I_M_Inventory> filter)
+	{
+		inventoryDAO.stream(filter)
+				.map(I_M_Inventory::getM_Inventory_ID)
+				.map(InventoryId::ofRepoId)
+				.forEach(this::recomputeForInventory);
+
+	}
+
+	public void recomputeForCostCollector(@NonNull final IQueryFilter<I_PP_Cost_Collector> filter)
+	{
+		ppCostCollectorDAO.stream(filter)
+				.forEach(this::recomputeForCostCollector);
+	}
+
+	public void recomputeForPPOrder(@NonNull final IQueryFilter<I_PP_Order> filter)
+	{
+		ppOrderDAO.stream(filter)
+				.map(I_PP_Order::getPP_Order_ID)
+				.map(PPOrderId::ofRepoId)
+				.forEach(this::recomputeForPPOrder);
+	}
+
+	private void recomputeForInvoice(@NonNull final InvoiceId invoiceId)
+	{
+		//dev-note: one trx per each document, to preserve the results of already successfully recomputed logs
+		trxManager.runInNewTrx(() -> invoiceDAO
+				.retrieveLines(invoiceId)
+				.forEach(line -> modularContractService.invokeWithModelForAllContractTypes(line, ModelAction.RECREATE_LOGS)));
+	}
+
+	private void recomputeForInOut(@NonNull final I_M_InOut inOut)
+	{
+		//dev-note: one trx per each document, to preserve the results of already successfully recomputed logs
+		trxManager.runInNewTrx(() -> inOutDAO
+				.retrieveAllLines(inOut)
+				.forEach(line -> modularContractService.invokeWithModelForAllContractTypes(line, ModelAction.RECREATE_LOGS)));
+	}
+
+	private void recomputeForOrder(@NonNull final I_C_Order order)
+	{
+		//dev-note: one trx per each document, to preserve the results of already successfully recomputed logs
+		trxManager.runInNewTrx(() -> orderDAO
+				.retrieveOrderLines(order)
+				.forEach(line -> modularContractService.invokeWithModelForAllContractTypes(line, ModelAction.RECREATE_LOGS)));
+	}
+
+	private void recomputeForFlatrate(@NonNull final I_C_Flatrate_Term term)
+	{
+		//dev-note: one trx per each document, to preserve the results of already successfully recomputed logs
+		trxManager.runInNewTrx(() -> modularContractService
```

Comment:
shall we also assert running out of trx?

Reply:
hmm... what do you think would be the problem if we are already in a trx when calling this?